add missing sri hashes

mosc9575 opened the pull request:
I added the sri hashes for 2.2.0 and 2.2.0. I made use of https://www.srihash.org/.

At the moment I added empty json files before version 0.8.0, because the calculator didn't return a valid answer and I was not able to guess the name or check any existence. 

Here is what I tried for 0.7.0
```
https://cdn.bokeh.org/bokeh/release/bokeh-0.7.0.js
```

- [ ] issues: fixes #13524

Commit:
add sri hashes for 2.2.0 and 2.2.2

Commit:
add empty json for sri hashes before 0.8.0

mosc9575 commented:
The empty json files result in failing unittests in `test_get_sri_hashes_for_version` like this:

> `FAILED tests/unit/bokeh/test_resources.py::TestSRIHashes::test_get_sri_hashes_for_version - AssertionError: assert 'bokeh-0.4.2.js' in {}`

Commit:
don not test for f"bokeh-{v}.js" before 0.8.0 in test_get_sri_hashes_for_version

bryevdv commented:
@mosc9575 here are all the bokeh 0.8 versions:

<img width="559" alt="Screenshot 2023-12-11 at 19 19 30" src="https://github.com/bokeh/bokeh/assets/1078448/4cc0cb46-4183-4dfe-89da-4dd5c481f1d9">

They are under `bokeh/release/` on the CDN:

https://cdn.bokeh.org/bokeh/release/bokeh-0.8.0.js

Due to unfortunate history, all the earlier versions are just at the top level of the CDN, e.g 

https://cdn.bokeh.org/bokeh-0.3.js

```
  | bokeh-0.3.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 1.0 MB | Standard
  | bokeh-0.3.min.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 478.5 KB | Standard
  | bokeh-0.4.1.css | css | July 19, 2019, 12:01:20 (UTC-07:00) | 107.5 KB | Standard
  | bokeh-0.4.1.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 937.7 KB | Standard
  | bokeh-0.4.1.min.css | css | July 19, 2019, 12:01:20 (UTC-07:00) | 90.2 KB | Standard
  | bokeh-0.4.1.min.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 414.9 KB | Standard
  | bokeh-0.4.2.css | css | July 19, 2019, 12:01:20 (UTC-07:00) | 107.2 KB | Standard
  | bokeh-0.4.2.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 951.2 KB | Standard
  | bokeh-0.4.2.min.css | css | July 19, 2019, 12:01:20 (UTC-07:00) | 90.0 KB | Standard
  | bokeh-0.4.2.min.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 421.5 KB | Standard
  | bokeh-0.4.3.css | css | July 19, 2019, 12:01:20 (UTC-07:00) | 135.2 KB | Standard
  | bokeh-0.4.3.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 1.2 MB | Standard
  | bokeh-0.4.3.min.css | css | July 19, 2019, 12:01:20 (UTC-07:00) | 117.5 KB | Standard
  | bokeh-0.4.3.min.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 571.9 KB | Standard
  | bokeh-0.4.4.css | css | July 19, 2019, 12:01:20 (UTC-07:00) | 135.2 KB | Standard
  | bokeh-0.4.4.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 1.2 MB | Standard
  | bokeh-0.4.4.min.css | css | July 19, 2019, 12:01:20 (UTC-07:00) | 117.5 KB | Standard
  | bokeh-0.4.4.min.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 571.9 KB | Standard
  | bokeh-0.4.css | css | July 19, 2019, 12:01:20 (UTC-07:00) | 106.5 KB | Standard
  | bokeh-0.4.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 909.3 KB | Standard
  | bokeh-0.4.min.css | css | July 19, 2019, 12:01:20 (UTC-07:00) | 89.3 KB | Standard
  | bokeh-0.4.min.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 401.7 KB | Standard
  | bokeh-0.5.0.css | css | July 19, 2019, 12:01:20 (UTC-07:00) | 188.4 KB | Standard
  | bokeh-0.5.0.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 2.1 MB | Standard
  | bokeh-0.5.0.min.css | css | July 19, 2019, 12:01:20 (UTC-07:00) | 164.7 KB | Standard
  | bokeh-0.5.0.min.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 1002.6 KB | Standard
  | bokeh-0.5.1.css | css | July 19, 2019, 12:01:20 (UTC-07:00) | 188.4 KB | Standard
  | bokeh-0.5.1.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 2.1 MB | Standard
  | bokeh-0.5.1.min.css | css | July 19, 2019, 12:01:20 (UTC-07:00) | 164.7 KB | Standard
  | bokeh-0.5.1.min.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 1002.3 KB | Standard
  | bokeh-0.5.2.css | css | July 19, 2019, 12:01:20 (UTC-07:00) | 188.4 KB | Standard
  | bokeh-0.5.2.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 2.1 MB | Standard
  | bokeh-0.5.2.min.css | css | July 19, 2019, 12:01:20 (UTC-07:00) | 164.7 KB | Standard
  | bokeh-0.5.2.min.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 1020.8 KB | Standard
  | bokeh-0.6.0.css | css | July 19, 2019, 12:01:20 (UTC-07:00) | 197.1 KB | Standard
  | bokeh-0.6.0.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 2.2 MB | Standard
  | bokeh-0.6.0.min.css | css | July 19, 2019, 12:01:20 (UTC-07:00) | 172.6 KB | Standard
  | bokeh-0.6.0.min.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 1.0 MB | Standard
  | bokeh-0.6.1.css | css | July 19, 2019, 12:01:20 (UTC-07:00) | 224.9 KB | Standard
  | bokeh-0.6.1.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 2.2 MB | Standard
  | bokeh-0.6.1.min.css | css | July 19, 2019, 12:01:20 (UTC-07:00) | 195.8 KB | Standard
  | bokeh-0.6.1.min.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 1.0 MB | Standard
  | bokeh-0.7.0.css | css | July 19, 2019, 12:01:20 (UTC-07:00) | 554.3 KB | Standard
  | bokeh-0.7.0.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 1.9 MB | Standard
  | bokeh-0.7.0.min.css | css | July 19, 2019, 12:01:20 (UTC-07:00) | 525.5 KB | Standard
  | bokeh-0.7.0.min.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 884.5 KB | Standard
  | bokeh-0.7.1.css | css | July 19, 2019, 12:01:20 (UTC-07:00) | 554.3 KB | Standard
  | bokeh-0.7.1.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 1.9 MB | Standard
  | bokeh-0.7.1.min.css | css | July 19, 2019, 12:01:20 (UTC-07:00) | 525.5 KB | Standard
  | bokeh-0.7.1.min.js | js | July 19, 2019, 12:01:20 (UTC-07:00) | 888.5 KB | Standard
```

bryevdv commented:
Bokeh versions before 0.3 are mostly lost to time, they are probably all but unbuildable and were not published to CDN. I guess some of them may still exist in the repo commit history (I forget when we stopped checking BokehJS in to the repo) or inside the PyPI packages (though I only see 0.2 there, nothing earlier)

mosc9575 commented:
@bryevdv Thanks for your help. Do you think it is good to add the hashes to the css files, too?

bryevdv commented:
> Do you think it is good to add the hashes to the css files, too?

If it's easy to do at the same time or automate, I guess it couldn't hurt. Just FYI adding these at all is really just a gold-star for "complete and correct". It's not impossible someone could want them for use in their own templates, but it's probably pretty unlikely.

Commit:
add hash for css and below version 0.8.0

Commit:
add hashes for version 0.4.3 and 2.2.1

Commit:
adapt 'test_get_sri_hashes_for_version'

mosc9575 commented:
I added the css for completeness before version 2.0.0, afterwards it doesn't exist anymore (see #9332). Version "0.3" and "0.4" have no trailing zero. Because of this I made changes to a test.

I also added a few missing files.

bryevdv commented on tests/unit/bokeh/test_resources.py:
```diff
@@ -66,11 +66,18 @@ def test_get_sri_hashes_for_version(self) -> None:
+                assert f"bokeh-{v_t}.min.js" in h
+            if "1" <= v:
+                assert f"bokeh-widgets-{v_t}.js" in h
+                assert f"bokeh-widgets-{v_t}.min.js" in h
```
Hi @mosc9575 I think this kind of logic indicates a test has grown over-complicated. There are a couple of things we can do, including splitting the "standard" and "weird" cases into their own tests, and using `parametrize` to avoid an explicit loop level. Given the arbitrary historical nature of these requirements, some comments are also probably helpful for posterity. 

What do you think of an approach like this (untested):

```python
from packaging.version import Version as V

ALL_VERSIONS = resources.get_all_sri_versions()

# very old Bokeh versions are inconsistent have have to be handled specially
STANDARD_VERSIONS = {v for v in ALL_VERSIONS if V(v) >= V("0.4")}
WIERD_VERSIONS = ALL_VERSIONS - STANDARD_VERSIONS

class TestSRIHashes:
   
   @pytest.mark.parametrize("v", STANDARD_VERSIONS)
    def test_get_sri_hashes_for_standard_versions(self, v) -> None:
        h = resources.get_sri_hashes_for_version(v)
        assert f"bokeh-{v}.js" in h
        assert f"bokeh-{v}.min.js" in h
        if v >= "1":
            assert f"bokeh-widgets-{v}.js" in h
            assert f"bokeh-widgets-{v}.min.js" in h

    @pytest.mark.parametrize("v", WIERD_VERSIONS)
    def test_get_sri_hashes_for_weird_versions(self) -> None:

        # 0.1.0 is tagged but untestable / no longer published
        if v == "0.1.0": 
            return 
        
        # other early versions omitted trailing .0 in filnames
        v = v.rstrip(".0")

        assert f"bokeh-{v}.js" in h
        assert f"bokeh-{v}.min.js" in h
```

bryevdv replied:
@bokeh/core there's also repo tags for `0.0.1` and `0.1.1` that don't have JSON files or releases on PyPI. I would not be opposed to deleting all the pre-0.2 tags (`0.0.1`, `0.1.0` and `0.1.1`) TBH, since I am not sure what useful purpose they serve at this point. That would simplify things here slightly as well, but we'd need to delete `0.1.0.rst` in the docs too. 

If we really want to keep `0.1.0.rst` and tag that's fine, but I'd still recommend deleting the other two tags.

tcmetzger replied:
I don't see any purpose in keeping any pre-0.2 tags either. Especially since it sounds like it would make resolving this issue easier, I'd say go for it!

bryevdv replied:
Thinking more I suppose we should keep 0.1.0 which is fine, but I will plan to delete the others soon unless there is objection. Then the only code changes for this PR are what I suggested above.

bryevdv commented:
I've deleted the 0.0.1 and 0.1.1 tags. Just for posterity, in case there really is ever some need to recall them for some obscure reason, I'll attach this screenshot with the commit hashes here:
<img width="404" alt="Screenshot 2023-12-18 at 10 25 29" src="https://github.com/bokeh/bokeh/assets/1078448/3027417d-0764-4d41-91f6-e1e8dd26f817">

Commit:
add test for WIERD_VERSIONS

Commit:
revert deletion of 'test_get_all_sri_versions_valid_format'

mosc9575 commented:
@bryevdv I updated the test based on your suggestions. The WIERD_VERSIONS are {'0.1.0', '0.2.0', '0.3.0', '0.4.0'}. I think this in fine now.

bryevdv commented:
Thank you @mosc9575 !
